### ISSUE TITLE
Deep Gamblers: Learning to Abstain with Portfolio Theory

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#40]
## 論文リンク
https://arxiv.org/abs/1907.00208

## 公開日（yyyy/mm/dd）
2019/06/29

## 概要
ポートフォリオ理論とディープラーニングの対応を考えて、そこから K クラス分類問題に1クラス追加して abstention までを含めた K+1 クラス分類問題の損失関数を提案。
具体的には、単勝の競馬でどの馬に賭けるかだけでなくお金を残して次のレースに備えるのに最適なポートフォリオを考え（この次のレースに備えるが abstention に対応）、doubling rate（あるポートフォリオを決めたときに stock の分布で積分してどれくらい資産が増加するか）が最適になるようなポートフォリオに対応する損失関数となる。
提案手法は、end-to-end で、サンプリング不要、要求される coverage 毎に再学習は必要ない、モデルアーキテクチャ変更不要、という利点があり、そのような利点がない過去手法よりも特に coverage がおおいい場合に良い結果。

**Comment by yoheikikuta** [comment]
selective classification は教師あり学習で rejection オプションのあるものを指すらしい。
この呼称は知らなかったが、いわゆる abstention（棄権）ありの分類問題で、自分が結構興味があるものでいくつかはチェックしてきたものだ。

NeurIPS 2019 accepted papers を眺めてたら目に留まり、ポートフォリオ理論との対応など自分には馴染みがないもので興味を惹かれたので読むことにした。

**Comment by yoheikikuta** [comment]
selective inference の定義は以下。
普通の分類モデル f 以外にも、confidence を測る一般に別のモデル g があって、g の出力が h 以上の場合には f の予測を採用してそうでなければ abstain するという枠組みである。

<p align="center">
  <img src="https://i.imgur.com/D5Q4pOH.png">
</p>

これを実現する方法は色々提案されている。

- Entropy Selection (ES)
  K class 分類モデルの出力を確率分布とみなしてエントロピーを計算して、その値が一定値以上なら abstain する。出力の要素が全て 1/K であればエントロピーは最大になるので、これは確かに望む用途には使えそうだ。分布が歪んでいさえすればよいので最適な感じはしないが。
- Softmax Response (SR)
  最もポピュラーだと思われるもので、softmax の出力が最大となる要素の値を confidence score とする。この confidence score が一定値より小さければ abstain する。最も自信があるものが十分自信があればそれを予測とする、というスタンスだ。単純だが結構上手くいく。
- Bayes Dropout (BD)
  bayesian に基づくもので、典型的には sampling を必要とするもの。自分では使ったことないな。
- Selective Net (SN)
  201901 に提案されたモデル。これを論文提案時の SOTA モデルとして扱っている。モデルの詳細には立ち入らないが、モデルアーキテクチャを変えたり達成したい coverage 毎に学習しなければならなかったりして使う際のコストは高い。

この論文では、ポートフォリオ理論からの類推で selective inference を可能にする loss function を定義し、それがお手軽でしかも性能が高いということを示していく。特徴の比較は以下。

<p align="center">
  <img src="https://i.imgur.com/MR0Ckqr.png">
</p>

**Comment by yoheikikuta** [comment]
ポートフォリオ理論とは何かを概説する。

まず、株式市場を考え、$` m `$ 個の株式を考える。
その確率変数を $` X = (X_1, ..., X_m) `$ とするが、こいつらは price relative という量になることに注意が必要。これは、一日の市場が開けた時間の価格を 1 としたときの市場が終わった時間の価格で、例えば 0.95 なら 5% マイナスというものである。
この $` X `$ は $` X ~ P(X) `$ からサンプルされるものとする。

次にポートフォリオを考える。
これは $` b = (b_1, ..., b_m) \ \text{where} \ b_i \geq 0, \sum_i b_i = 1 `$ で、どの株式をどれくらいの割合で持つかを定めているものである。総和が 1 なので（株式に使おうと思っている）全財産を使っていると思えば良い。
ここでは手数料など諸々の要素は考慮してないかなり単純化したものになっている。

以上を用いて wealth relative $` S = b^T X `$ が定義され、これは一日の開始と比べて終了時にどれくらい富が増しているかを示すものになっている。
この wealth relative を最大化するようにポートフォリオを組みたいわけで、それを何かしらの指針に基づいて数学的に解き明かしていくのがポートフォリオ理論である。

特に重要となる量に以下で定義される doubling rate がある。
これはポートフォリオを定めたとき、wealth relative の log_2 の期待値をとったものになっている。例えばこの値が 1 ならば期待値として wealth relative が 2 (つまり一日で資産が2倍) になるようなポートフォリオということになる。無論確率分布 $` P(X) `$ など分かるものではないが。
上で何かしらの指針に基づいて、と言ったが一つの例がこの doubling rate を最大化するように、という塩梅だ。

<p align="center">
  <img src="https://i.imgur.com/zSQb6xC.png">
</p>

だいぶ簡単化しているけどこの論文を理解するのに必要なのはこれくらい。
ここから問題をより限定していって分類問題とうまく対応するもの（単勝レースの競馬）を考えていくことになる。

**Comment by yoheikikuta** [comment]
この論文では分類問題を解くことをターゲットにするので、ポートフォリオはどれか単独の対象に全部賭けるのが対応するものになる。
それは株式市場というより、単勝競馬レースだ！ということで単勝競馬レースを考えていく。複勝とかワイドとかはここでは考えない（例えばマルチラベル分類を考えるとかだったら必要になるかもしれない）。

こうすると一回のレースでの wealth relative は $` S =(x(j)) = b_j o_j `$ となる。これた単に $` x(j) = \text{(j 成分だけ 1 であとは 0)} `$ としているから。

$` n `$ 回レースを繰り返すことを考える。レースの順番は関係ないので $` S_n = \prod_a S(\mathbf{x}_a) `$ のように単なる積で書ける（いままでサボったけどここだけ $` x `$ を太文字で書いた。この添字 a はレースの index で $` \mathbf{x}_a `$ は馬の数の次元で一要素だけ 1 をとるもの）。
この順番に依らないというのが、分類問題で loss を計算する際に順番に依らないのでバッチ処理が可能であることに対応する。

単勝競馬レースの場合に doubling rate を書き換えると以下のようになる。離散化して馬 $` i `$ が勝つ確率を $` p_i `$ としているというだけ。

<p align="center">
  <img src="https://i.imgur.com/uTLFYjc.png">
</p>

この形を見ると、$` o_i = 1 `$ として $` b_i `$ は softmax の出力値とすれば、これは分類でよく使う cross entropy (の overall sign を変えたもの)となることが分かる。

この結果ではポートフォリオを持ち出してきたご利益はまだ分からないので、もう一歩進めてみる。
軍資金を準備して競馬に行ったとき、毎レース毎に全額賭ける人がいるだろうか？いるかもしれないが、どの馬が勝つと思っているかの自信に応じて、今回は賭け金を減らして次のレースに備えておく、という方がより現実的な戦略だろう。
それを取り込むには、馬の頭数を $` m `$ としたときに wealth relative を $` S(x(j)) = b_j o_j + b_{m+1} `$ のように項を追加すればよい（追加した上で、ポートフォリオの総和を 1 になるようにする）。
これを maximize すればよい、ということで以下を解くことになる。ちょっと notation が分かりづらいが、b を動かして右辺を最大化する。

<p align="center">
  <img src="https://i.imgur.com/chbcqR4.png">
</p>

シンプルだがこれが重要で、まさに abstention に対応するものが導入されたので、あとはこれを分類問題として読み替えていくことになる。
ここまでで、ポートフォリオ理論とディープラーニングの対応は以下のようになることが分かった。

<p align="center">
  <img src="https://i.imgur.com/7MsmACB.png">
</p>

**Comment by yoheikikuta** [comment]
先ほどの目的関数を分類問題に読みかえよう。

$` m `$クラスの分類をするモデルを $` f `$ として、$` f(x) `$ は確率値として解釈できるように規格化されているものとする。$` P(j|x) = f(x)_j `$ と書く。このモデルはパラメタ $` w `$ で特徴付けられるものとする。
これは普通の分類モデルなので、abstention を取り入れて $` m+1 `$ クラスの分類問題に拡張する。
ある mini-batch $` B `$ に対する目的関数を、先ほどの結果を考慮して構築したものが以下。

<p align="center">
  <img src="https://i.imgur.com/SVh9j9R.png">
</p>

$` o `$ は定数になっている。競馬の場合はある馬が勝った場合の return だったので馬毎に異なるものだったが、分類問題は一番シンプルなケースはどのクラスが正解でも違いがないので、こういう扱いをしているわけだ。
そしたら $` o = 1 `$ にしとけばいいのでは、という話があるが、これはターゲットにしている $` m `$ クラスの場合と abstention の場合の重みをつけるために存在している。これは hyperparameter となる。

この $` o `$ は $` 1 < o < m `$ のみで意味を持つ。
これはちゃんと証明もできるが、1 より小さかったら $` f(x)_{m+1}  = 1 `$ に寄せればいいし、$` m `$ より大きかったら $` 0 `$ とすればいいことはすぐに分かる。

ということでこれが提案手法である。$` m + 1 `$ クラス分類にしてこの目的関数で学習せよ、ということで結果としてはかなりシンプル。

**Comment by yoheikikuta** [comment]
実験に入る前に、単勝競馬レースの場合に情報論的な議論を適用している。
内容としては、

- optimal な doubling rate が存在してそれを実現するポートフォリオを示す
- side information X を使って単勝競馬レース Y を考えるとき、doubling rate の増加分は X, Y の相互情報量で bound されることを示す

というものになっている。
証明は Appendix に載っていて特に難しいところはないのでここでは割愛しておく。

二つ目は少し補足や解釈のしがいがある。
まず side information とは、競馬で言えば過去の戦績とかパドックで見た馬の状態とかに当たるものだ。こういう情報に基づいてどの馬が勝つかという予測を当たりやすくしたいわけで、それは相互情報量で定められる。
直感的には、side information が持つ情報からどれくらいレース結果が予測できるかということなので違和感はないだろう。エントロピーの言葉で言えば、もともとのレース結果が持つ不確実性が、side information が与えれることでどれくらい減少するかということで、reasonable な感じはする。
ディープラーニングの場合はこれは取りも直さず入力のデータに対応する。画像ならピクセル値によってラベルを予測するという話に過ぎない。

これを以って論文では

> This means that the neural network, during training, will have to
maximize the mutual information between the prediction and the true label of the sample.

と述べているがこれはちと分からん。
相互情報量は入力とラベルの話なので、予測性能は入力データとラベルの相互情報量で bound されるということで、それは make sense だ。しかし prediction となると、X を given にした prediction とラベルの相互情報量を最大化すべし、と言うべきだろう。

この辺は information bottleneck 的な議論とも密接に関係しているので、そういう観点からもっとポートフォリオ理論との関連性を深めていくのは面白いのかもしれない。

**Comment by yoheikikuta** [comment]
ということで実験。

まずは人工的に生成したデータから。
2-layer, 50 hidden, tanh activation のネットワークで、2d のガウス分布から生成されたデータを分類する問題。これでは説明が不十分だが結果の図を見れば想像できるので結果を貼る。黄色は abstention で、下の図の緑色はテスト時だけに準備した学習時には出会ってないクラスと考えられるデータである。

<p align="center">
  <img src="https://i.imgur.com/4QNA1ZZ.png">
</p>

ということでうまく動いている。
entropy based でも動いてはいるが、ちゃんと予測できるはずのところも abstention してしまっているので、提案手法の方が適切に abstention できている。

次に MNIST での実験。
4-layer CNN を使用。abstention の確率が高いものを entropy based と比べたものが以下。
entropy based よりも人間から見ても分からないようなものが並んでいる感じはする。

<p align="center">
  <img src="https://i.imgur.com/a1ykDdE.png">
</p>

次に MNIST の 9 を少しずつ回転させていって予測した結果。
これがなかなか見せ方も面白くて良かった。180° 回せば 6 になるので再び予測できるが、途中では予測できないようなものが出てきてるはずで、それが一程度ちゃんと abstention できることが示されている。
途中 5 と予測してしまっているのがちょっと惜しい（まあ 0~9 のどれに一番近いかと問われれば 5 ではある）。

<p align="center">
  <img src="https://i.imgur.com/2SPUukh.png">
</p>

**Comment by yoheikikuta** [comment]
ここまではモデルの特徴を掴むための実験で、以降が既存手法との比較。
SVHN, CIFAR10, Cats vs. Dogs で実験しているが、傾向は概ね変わらないので CIFAR10 のもののみ載せておく。モデルは VGG で使ってるデータと共に SelctiveNet の論文に合わせている。

実験としては、abstention が存在するので、テストデータのうちどれくらいを abstain して、ちゃんと予測したものがどれくらいエラー率が低いのかで比べている。例えば coverage 0.95 ならば、テストデータのうち 5% は abstain されていて、残りの 95% でのエラー率を比べることになる。

<p align="center">
  <img src="https://i.imgur.com/GITaSuC.png">
</p>

o の値は 0.2 毎に grid search をして得られた最良のものを使用していて、標準偏差は +-0.2 の値でも計算して計算しているとのこと。o の値に対する sensitivity のようなものだと思っておけばよいだろう。
single best model というのは o の値が最も良いものを採用してそれで全ての coverage の結果を出したもので、best per coverage は coverage の値毎に最適な o を求めたものになっている。後者は最初に言った retrain なしの利点を捨てているわけだが、可能な限り tune したら最高性能はどれくらいかという意味で出してるんだろう。

結果としては coverage が大きいところでは既存手法を outperform しているということでなかなか悪くない。coverage を下げていくと必ずしも最善の手法という感じではなくなってくるが、このお手軽さでこの結果ならなかなかのコスパの良さだと思う。

**Comment by yoheikikuta** [comment]
特徴量がより separable になるという結果も載せている。
ただこれはなんでこうなるかとか妥当な説明がつけられているわけではない。

<p align="center">
  <img src="https://i.imgur.com/VvwQkOq.png">
</p>

**Comment by yoheikikuta** [comment]
ということで一通り読んだ。
ポートフォリオ理論との対応というのは面白いし、提案手法としてシンプルだし結果も悪くない。
一番初めのごく簡単な例をやったという感じなので、界隈に興味がある人がいれば掘っていけるところは色々ありそうな感じだ。
